allow executable (+x) non-script inventory files.

danthedeckie opened the pull request:
This fixes #10068 without breaking backwards compatibility.

Sometimes having an ansible directory on a network share means it mounts +x automatically, with no easy way to get around it.

Commit:
allow executable (+x) non-script inventory files.

abadger commented:
Can we autodetect this?  If executable, try to execute the file as a script and if that fails, try to read it as an inventory file?

abadger commented:
Or vice-versa, try to read the file as an inventory file.  If that fails to parse and the file is executable, then try to execute it as an inventory script.

srgvg commented:
Actually, it depends. If the inventory points directly to a script, and  the file is executeable, it will also check for the presence of a shebang. It will fail is no shebang is present. One could change this to fallback to ini parsing. However, if the inventory points to a directory, that shebang check is not performed, and only the executeable flag is checked. The former is inventory/**init**.py, the latter is of course the evilsh inventory_dir.py, which reimplements some of the former code... One of the things for after v2, me thinks.

danthedeckie commented:
shebang detection does limit you to scripting languages.  I could see people who use exclusively golang (for instance) wanting their inventory program also in the same language.  Sure, you could use a wrapper script...

bcoca commented:
I'm almost tempted to just try reading it as the static format if the execution fails for any reason, then just give back an error that we could neither execute nor parse the file.

srgvg commented:
@bcoca which is what I said. Only you have two code paths to take care of this now.

bcoca commented:
2 places which should call the same function .... we have too many double implementations of things already.

srgvg commented:
Hear! Hear!

Commit:
Attempt to run executable scripts, and if that fails, try parsing as INI,
and if that fails, give an appropriate error message.  Also minor refactoring
to make those changes in a cleaner manner.

danthedeckie commented:
How about this?  It's a more complex change, a bit of refactoring, but I think implements the expected behaviour, without (hopefully) breaking any current systems, and still remaining reasonably flexible.

danthedeckie commented:
As a side-topic / question, is there a reason more custom Exceptions aren't used?  They make unit tests a heck of a lot more robust, and easier to write.  Or is this a thing for v2 / later on?

bcoca commented:
the custom exceptions is something we introduced much later, we are moving to that as we touch code.

danthedeckie commented:
_bump_ does it still needs_revision?  Is this a good solution?

cheesegrits commented:
_bump_ can this issue get some love? Still a nasty problem for us Winders Vagrant users.

amenonsen commented:
@danthedeckie @cheesegrits can you try out PR #11643 to see if it solves your problem? If so, this PR can be closed.

abadger commented:
Okay, we think that this was fixed by #11643 which was merged a few days ago.  I'll go ahead and close this.  If there's further code that's needed to get his working, please open a new issue/pull request and we can get it sorted out.

# Closing This Ticket

Hi!

We believe recent commits (likely detailed above) should resolve this question or problem for you.

This will also be included in the next major release.

If you continue seeing any problems related to this issue, or if you have any further questions, please let us know by stopping by one of the two mailing lists, as appropriate:
- https://groups.google.com/forum/#!forum/ansible-project - for user questions, tips, and tricks
- https://groups.google.com/forum/#!forum/ansible-devel - for strategy, future planning, and questions about writing code

Because this project is very active, we're unlikely to see comments made on closed tickets, but the mailing list is a great way to ask questions, or post if you don't think this particular
issue is resolved.

Thank you!
